### DISCUSSION TITLE
Ax = B wrong solution when A contains a row of zeros

### DISCUSSION THREAD

**Discussion by eladr07** [mathnet/mathnet-numerics#822] [Q&A] [answered]
Hi,

I am using MathNet.Numerics v4.7.0.

I try to solve Ax = B (for x) when A is a 4 by 6 matrix with the following value array (note that the first row is all zeros - indices 0,7,13,19):

    [0]: 0
    [1]: 1.06261732
    [2]: 0
    [3]: 0.00038268
    [4]: 0
    [5]: 1.063
    [6]: 0
    [7]: 0
    [8]: 3.335926901
    [9]: 0.020869227
    [10]: 0.104069263
    [11]: 3.4609
    [12]: 0
    [13]: 0
    [14]: 0
    [15]: 1.0538
    [16]: 0
    [17]: 1.0538
    [18]: 0
    [19]: 0
    [20]: 0
    [21]: 0.000663424
    [22]: 0.933717888
    [23]: 0.9344

And B is a 6-vector with the following values (note that first element is non-zero):

    [0]: -147339.22693314467
    [1]: 90907.703152322269
    [2]: -71858.548209015382
    [3]: 67094.2200240139
    [4]: 61199.2777493189
    [5]: 0

And A.Solve(B) returns a solution which is obviously incorrect.

What am I missing?

Thanks!

**Comment by cdrnet** [comment, marked as answer]
`Solve` is supposed to return the solution which minimizes the squared errors i.e. `|Ax - b|`. The returned vector looks ok to me. Is there another solution with a lower least-squares error?

Note that the error is not necessarily zero. In your particular problem there exists no solution with zero error, since there is no solution for the equation `0 * t = -147339.22693314467` corresponding to the first row of your 6x4 matrix and the first element of your 6-vector.

**Comment by eladr07** [comment]
I understand now, thanks for your explanation.

Is there any way to determine how "good" is the solution? Currently I'm looking at determining it from a business POV and then looking at 
```
var error = A
                .Multiply(x)
                .Subtract(B)
                .Norm(2.0);
```

**Comment by cdrnet** [comment]
For a start, you could check whether the rank of the augmented matrix A|b is greater than the rank of A. If so, the system is inconsistent and there exists no solution - like in your example, where the former is 5 and the latter 4.

```
if(A.Append(b.ToColumnMatrix()).Rank() > A.Rank())
{
    // inconsistent, i.e. there is no solution
}
```
